File: src/adoption/gaps.py
"""Gap analysis — compare project state against template requirements.

Detects missing documents, stub documents, and known deficiencies.
Generates GAPS.md from analysis.
"""

from __future__ import annotations

import os
from dataclasses import dataclass, field
from typing import Literal

from .inventory import Inventory
from .templates import Template, TemplateSet


GapSeverity = Literal["critical", "high", "medium", "low"]
GapStatus = Literal["open", "in-progress", "closed", "not-applicable"]


@dataclass
class Gap:
    """A single gap between project state and template requirements."""
    template_name: str
    template_file: str
    severity: GapSeverity
    status: GapStatus = "open"
    reason: str = ""
    justification: str = ""  # For not-applicable items


@dataclass
class GapAnalysis:
    """Complete gap analysis result."""
    project_dir: str
    gaps: list[Gap] = field(default_factory=list)

    @property
    def critical(self) -> list[Gap]:
        return [g for g in self.gaps if g.severity == "critical" and g.status == "open"]

    @property
    def open_gaps(self) -> list[Gap]:
        return [g for g in self.gaps if g.status == "open"]

    @property
    def by_severity(self) -> dict[str, list[Gap]]:
        result: dict[str, list[Gap]] = {}
        for g in self.gaps:
            if g.status != "open":
                continue
            result.setdefault(g.severity, []).append(g)
        return result

    def to_markdown(self) -> str:
        """Generate GAPS.md content."""
        lines = ["# GAPS — Gap Analysis", ""]
        lines.append(f"**Total gaps:** {len(self.open_gaps)} open")
        lines.append(f"**Critical:** {len(self.critical)}")
        lines.append("")

        for severity in ("critical", "high", "medium", "low"):
            gaps_at_level = [g for g in self.gaps if g.severity == severity and g.status == "open"]
            if gaps_at_level:
                lines.append(f"## {severity.title()}")
                lines.append("")
                for gap in gaps_at_level:
                    lines.append(f"- [ ] **{gap.template_name}** (`{gap.template_file}`)")
                    if gap.reason:
                        lines.append(f"  - {gap.reason}")
                lines.append("")

        # Not-applicable items
        na_items = [g for g in self.gaps if g.status == "not-applicable"]
        if na_items:
            lines.append("## Not Applicable")
            lines.append("")
            for gap in na_items:
                lines.append(f"- **{gap.template_name}**: {gap.justification}")
            lines.append("")

        return "\n".join(lines)


TODO_MARKERS = ["TODO", "FIXME", "TBD", "PLACEHOLDER", "STUB"]


def _is_stub(filepath: str) -> bool:
    """Check if a file is a stub (contains TODO markers, very short)."""
    try:
        with open(filepath) as f:
            content = f.read()
        if len(content.strip()) < 50:
            return True
        upper = content.upper()
        return any(marker in upper for marker in TODO_MARKERS)
    except (OSError, UnicodeDecodeError):
        return False


def analyze_gaps(
    project_dir: str,
    inventory: Inventory,
    template_set: TemplateSet,
) -> GapAnalysis:
    """Compare project inventory against required templates.

    Checks:
    1. Missing documents (template required, file doesn't exist)
    2. Stub documents (file exists but has TODO markers)
    3. Severity mapping: R=critical, P=high, M=medium, O=low
    """
    analysis = GapAnalysis(project_dir=project_dir)
    existing_files = {item.path.lower() for item in inventory.items}

    severity_map = {
        "R": "critical",
        "P": "high",
        "M": "medium",
        "O": "low",
    }

    for template in template_set.templates:
        template_lower = template.filename.lower()

        # Check if file exists
        found = template_lower in existing_files
        if not found:
            # Also check without docs/ prefix
            alt = template_lower.replace("docs/", "")
            found = alt in existing_files

        if not found:
            analysis.gaps.append(Gap(
                template_name=template.name,
                template_file=template.filename,
                severity=severity_map.get(template.requirement, "low"),
                reason="Document does not exist",
            ))
            continue

        # Check if it's a stub
        full_path = os.path.join(project_dir, template.filename)
        if os.path.exists(full_path) and _is_stub(full_path):
            analysis.gaps.append(Gap(
                template_name=template.name,
                template_file=template.filename,
                severity=severity_map.get(template.requirement, "low"),
                reason="Document is a stub (contains TODO markers or is too short)",
            ))

    return analysis

File: rust/src/adoption/gaps.rs
//! Gap analysis — compare project state against template requirements.
//!
//! Detects missing documents, stub documents, and known deficiencies.
//! Generates GAPS.md from analysis.

use std::collections::HashSet;
use std::fs;
use std::path::Path;

use serde::{Deserialize, Serialize};

use super::inventory::Inventory;
use super::templates::{RequirementLevel, TemplateSet};

/// Gap severity.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum GapSeverity {
    Critical,
    High,
    Medium,
    Low,
}

impl GapSeverity {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Critical => "critical",
            Self::High => "high",
            Self::Medium => "medium",
            Self::Low => "low",
        }
    }
}

/// Gap status.
#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "kebab-case")]
pub enum GapStatus {
    #[default]
    Open,
    InProgress,
    Closed,
    NotApplicable,
}

/// A single gap between project state and template requirements.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Gap {
    pub template_name: String,
    pub template_file: String,
    pub severity: GapSeverity,
    pub status: GapStatus,
    pub reason: String,
    pub justification: String,
}

/// Complete gap analysis result.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GapAnalysis {
    pub project_dir: String,
    pub gaps: Vec<Gap>,
}

impl GapAnalysis {
    pub fn critical(&self) -> Vec<&Gap> {
        self.gaps
            .iter()
            .filter(|g| g.severity == GapSeverity::Critical && g.status == GapStatus::Open)
            .collect()
    }

    pub fn open_gaps(&self) -> Vec<&Gap> {
        self.gaps
            .iter()
            .filter(|g| g.status == GapStatus::Open)
            .collect()
    }

    pub fn by_severity(&self) -> std::collections::HashMap<String, Vec<&Gap>> {
        let mut result: std::collections::HashMap<String, Vec<&Gap>> = std::collections::HashMap::new();
        for g in &self.gaps {
            if g.status != GapStatus::Open {
                continue;
            }
            result
                .entry(g.severity.as_str().to_string())
                .or_default()
                .push(g);
        }
        result
    }

    /// Generate GAPS.md content.
    pub fn to_markdown(&self) -> String {
        let mut lines = vec![
            "# GAPS — Gap Analysis".to_string(),
            String::new(),
            format!("**Total gaps:** {} open", self.open_gaps().len()),
            format!("**Critical:** {}", self.critical().len()),
            String::new(),
        ];

        for severity in ["critical", "high", "medium", "low"] {
            let gaps_at_level: Vec<&Gap> = self
                .gaps
                .iter()
                .filter(|g| g.severity.as_str() == severity && g.status == GapStatus::Open)
                .collect();
            if !gaps_at_level.is_empty() {
                let title = severity[..1].to_uppercase() + &severity[1..];
                lines.push(format!("## {title}"));
                lines.push(String::new());
                for gap in &gaps_at_level {
                    lines.push(format!(
                        "- [ ] **{}** (`{}`)",
                        gap.template_name, gap.template_file
                    ));
                    if !gap.reason.is_empty() {
                        lines.push(format!("  - {}", gap.reason));
                    }
                }
                lines.push(String::new());
            }
        }

        // Not-applicable items
        let na_items: Vec<&Gap> = self
            .gaps
            .iter()
            .filter(|g| g.status == GapStatus::NotApplicable)
            .collect();
        if !na_items.is_empty() {
            lines.push("## Not Applicable".to_string());
            lines.push(String::new());
            for gap in &na_items {
                lines.push(format!("- **{}**: {}", gap.template_name, gap.justification));
            }
            lines.push(String::new());
        }

        lines.join("\n")
    }
}

const TODO_MARKERS: &[&str] = &["TODO", "FIXME", "TBD", "PLACEHOLDER", "STUB"];

fn is_stub(filepath: &str) -> bool {
    let content = match fs::read_to_string(filepath) {
        Ok(c) => c,
        Err(_) => return false,
    };
    if content.trim().len() < 50 {
        return true;
    }
    let upper = content.to_uppercase();
    TODO_MARKERS.iter().any(|marker| upper.contains(marker))
}

fn severity_from_requirement(req: &RequirementLevel) -> GapSeverity {
    match req {
        RequirementLevel::R => GapSeverity::Critical,
        RequirementLevel::P => GapSeverity::High,
        RequirementLevel::M => GapSeverity::Medium,
        RequirementLevel::O => GapSeverity::Low,
    }
}

/// Compare project inventory against required templates.
pub fn analyze_gaps(project_dir: &str, inventory: &Inventory, template_set: &TemplateSet) -> GapAnalysis {
    let mut analysis = GapAnalysis {
        project_dir: project_dir.to_string(),
        gaps: Vec::new(),
    };

    let existing_files: HashSet<String> = inventory
        .items
        .iter()
        .map(|item| item.path.to_lowercase())
        .collect();

    for template in &template_set.templates {
        let template_lower = template.filename.to_lowercase();

        let mut found = existing_files.contains(&template_lower);
        if !found {
            // Also check without docs/ prefix
            let alt = template_lower.replace("docs/", "");
            found = existing_files.contains(&alt);
        }

        if !found {
            analysis.gaps.push(Gap {
                template_name: template.name.clone(),
                template_file: template.filename.clone(),
                severity: severity_from_requirement(&template.requirement),
                status: GapStatus::Open,
                reason: "Document does not exist".to_string(),
                justification: String::new(),
            });
            continue;
        }

        // Check if it's a stub
        let full_path = Path::new(project_dir).join(&template.filename);
        if full_path.exists() && is_stub(&full_path.to_string_lossy()) {
            analysis.gaps.push(Gap {
                template_name: template.name.clone(),
                template_file: template.filename.clone(),
                severity: severity_from_requirement(&template.requirement),
                status: GapStatus::Open,
                reason: "Document is a stub (contains TODO markers or is too short)".to_string(),
                justification: String::new(),
            });
        }
    }

    analysis
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::adoption::inventory::{Inventory, InventoryItem};
    use crate::adoption::templates::{Template, TemplateSet};

    #[test]
    fn test_analyze_gaps_missing_document() {
        let inventory = Inventory {
            project_dir: "/test".to_string(),
            items: vec![],
        };
        let ts = TemplateSet {
            templates: vec![Template::new(
                "code",
                "README",
                "Project overview",
                "README.md",
                RequirementLevel::R,
            )],
        };
        let result = analyze_gaps("/test", &inventory, &ts);
        assert_eq!(result.gaps.len(), 1);
        assert_eq!(result.gaps[0].severity, GapSeverity::Critical);
        assert_eq!(result.gaps[0].reason, "Document does not exist");
    }

    #[test]
    fn test_analyze_gaps_existing_document() {
        let inventory = Inventory {
            project_dir: "/test".to_string(),
            items: vec![InventoryItem {
                path: "README.md".to_string(),
                format: "markdown".to_string(),
                size_bytes: 0,
                last_modified: 0.0,
                quality: "usable".to_string(),
                notes: String::new(),
            }],
        };
        let ts = TemplateSet {
            templates: vec![Template::new(
                "code",
                "README",
                "Overview",
                "README.md",
                RequirementLevel::R,
            )],
        };
        // The file doesn't actually exist on disk, so is_stub won't trigger
        let result = analyze_gaps("/test", &inventory, &ts);
        assert!(result.gaps.is_empty());
    }

    #[test]
    fn test_analyze_gaps_stub_detection() {
        let dir = tempfile::tempdir().unwrap();
        let readme_path = dir.path().join("README.md");
        std::fs::write(&readme_path, "# TODO: Fill this in").unwrap();

        let inventory = Inventory {
            project_dir: dir.path().to_str().unwrap().to_string(),
            items: vec![InventoryItem {
                path: "README.md".to_string(),
                format: "markdown".to_string(),
                size_bytes: 21,
                last_modified: 0.0,
                quality: "usable".to_string(),
                notes: String::new(),
            }],
        };
        let ts = TemplateSet {
            templates: vec![Template::new(
                "code",
                "README",
                "Overview",
                "README.md",
                RequirementLevel::R,
            )],
        };
        let result = analyze_gaps(dir.path().to_str().unwrap(), &inventory, &ts);
        assert_eq!(result.gaps.len(), 1);
        assert!(result.gaps[0].reason.contains("stub"));
    }

    #[test]
    fn test_gap_analysis_to_markdown() {
        let analysis = GapAnalysis {
            project_dir: "/test".to_string(),
            gaps: vec![
                Gap {
                    template_name: "README".to_string(),
                    template_file: "README.md".to_string(),
                    severity: GapSeverity::Critical,
                    status: GapStatus::Open,
                    reason: "Missing".to_string(),
                    justification: String::new(),
                },
                Gap {
                    template_name: "LICENSE".to_string(),
                    template_file: "LICENSE".to_string(),
                    severity: GapSeverity::High,
                    status: GapStatus::Open,
                    reason: "Missing".to_string(),
                    justification: String::new(),
                },
            ],
        };
        let md = analysis.to_markdown();
        assert!(md.contains("# GAPS"));
        assert!(md.contains("**Total gaps:** 2 open"));
        assert!(md.contains("## Critical"));
        assert!(md.contains("README"));
    }

    #[test]
    fn test_gap_analysis_critical_filter() {
        let analysis = GapAnalysis {
            project_dir: "/test".to_string(),
            gaps: vec![
                Gap {
                    template_name: "A".to_string(),
                    template_file: "a".to_string(),
                    severity: GapSeverity::Critical,
                    status: GapStatus::Open,
                    reason: String::new(),
                    justification: String::new(),
                },
                Gap {
                    template_name: "B".to_string(),
                    template_file: "b".to_string(),
                    severity: GapSeverity::Low,
                    status: GapStatus::Open,
                    reason: String::new(),
                    justification: String::new(),
                },
            ],
        };
        assert_eq!(analysis.critical().len(), 1);
        assert_eq!(analysis.open_gaps().len(), 2);
    }
}
